# SQL

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.
Компания решила быть на волне и купила крупный сервис для чтения книг по подписке. В базе данных содержится информация о книгах, издательствах, авторах, а также пользовательские обзоры книг.


Цель работы: анализ базы данных. Для этого понадобится:
1. Посчитать, сколько книг вышло после 1 января 2000 года;
2. Для каждой книги посчитать количество обзоров и среднюю оценку;
3. Определить издательство, которое выпустило наибольшее число книг толще 50 страниц;
4. Определить автора с самой высокой средней оценкой книг, учитывая только книги с 50 и более оценками;
5. Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок;
6. Cформулировать основные выводы.


<a id='Section_0'></a>
### Оглавление

[1. Открытие файла с данными и изучение общей информации](#Section_1)

[2. Количество книг, вышедших после 1 января 2000 года](#Section_2)

[3. Количество обзоров и средняя оценка](#Section_3)

[4. Издательство, которое выпустило наибольшее число книг толще 50 страниц](#Section_4)

[5. Автор с самой высокой средней оценкой книг](#Section_5)

[6. Среднее количество обзоров от пользователей, которые поставили больше 50 оценок](#Section_6)

[7. Выводы](#Section_7)

<a id='Section_1'></a>
## 1. Открытие файла с данными и изучение общей информации

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine

db_config = {'user': 'praktikum_student', # имя пользователя
            'pwd': 'Sdf4$2;d-d30pp', # пароль
            'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
            'port': 6432, # порт подключения
            'db': 'data-analyst-final-project-db'} # название базы данных


connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                db_config['pwd'],
                                                db_config['host'],
                                                db_config['port'],
                                                db_config['db'])

engine = create_engine(connection_string)  

### Таблица books

In [2]:
query = '''
            SELECT * FROM books

        '''
books = pd.io.sql.read_sql(query, con = engine)
display(books.head())
books.info()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


### Таблица authors

In [3]:
query = '''
            SELECT * FROM authors

        '''
authors = pd.io.sql.read_sql(query, con = engine)
display(authors.head())
authors.info()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


### Таблица publishers

In [4]:
query = '''
            SELECT * FROM publishers

        '''
publishers = pd.io.sql.read_sql(query, con = engine)
display(publishers.head())
publishers.info()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


### Таблица ratings

In [5]:
query = '''
            SELECT * FROM ratings

        '''
ratings = pd.io.sql.read_sql(query, con = engine)
display(ratings.head())
ratings.info()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


### Таблица reviews

In [6]:
query = '''
            SELECT * FROM reviews

        '''
reviews = pd.io.sql.read_sql(query, con = engine)
display(reviews.head())
reviews.info()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


Все таблицы загружены. Пропусков нет. Столбцы в нижнем регистре.

[Вернуться к оглавлению](#Section_0)

<a id='Section_2'></a>
## 2. Количество книг, вышедших после 1 января 2000 года

In [7]:
query = '''
SELECT 
    COUNT (book_id) AS books_cnt
FROM 
    books
WHERE 
    publication_date >= '2000-01-01'

'''
pd.io.sql.read_sql(query, con = engine)

books_cnt
0        821

После 1 января 2000 года вышла 821 книга.

<a id='Section_3'></a>
## 3. Количество обзоров и средняя оценка

In [8]:
query = '''
SELECT 
    books.book_id,
    books.title,
    COUNT(DISTINCT reviews.review_id) AS reviews_cnt,
    AVG(ratings.rating) AS avg_ratings      
FROM 
    books
    LEFT JOIN ratings ON ratings.book_id = books.book_id 
    LEFT JOIN reviews ON reviews.book_id = books.book_id 
GROUP BY
    books.book_id
ORDER BY
    reviews_cnt DESC

       '''
avg = pd.io.sql.read_sql(query, con = engine)
display(avg.head())
avg.tail(10)

book_id                                              title  reviews_cnt  \
0      948                            Twilight (Twilight  #1)            7   
1      963                                Water for Elephants            6   
2      734                                   The Glass Castle            6   
3      302  Harry Potter and the Prisoner of Azkaban (Harr...            6   
4      695  The Curious Incident of the Dog in the Night-Time            6   

   avg_ratings  
0     3.662500  
1     3.977273  
2     4.206897  
3     4.414634  
4     4.081081

book_id                                              title  reviews_cnt  \
990      413                                         Lysistrata            1   
991      599  Stone of Farewell (Memory  Sorrow  and Thorn  #2)            1   
992      535             Ranma ½  Vol. 1 (Ranma ½ (US 2nd)  #1)            1   
993      510                      Phantom (Sword of Truth  #10)            1   
994      387                               Leonardo's Notebooks            0   
995       83    Anne Rice's The Vampire Lestat: A Graphic Novel            0   
996      808                            The Natural Way to Draw            0   
997      672   The Cat in the Hat and Other Dr. Seuss Favorites            0   
998      221                          Essential Tales and Poems            0   
999      191  Disney's Beauty and the Beast (A Little Golden...            0   

     avg_ratings  
990     4.000000  
991     5.000000  
992     4.000000  
993     3.500000  
994     4.000000  
995     3.666667  
996     3.000000  
997     5.000000  
998     4.000000  
999     4.000000

Самое большое количество обзоров у книги Twilight (Twilight #1) со средней оценкой 3,66. Есть 6 книг, которые не имеют пользовательских обзоров.

[Вернуться к оглавлению](#Section_0)

<a id='Section_4'></a>
## 4. Издательство, которое выпустило наибольшее число книг толще 50 страниц

In [9]:
query = '''
SELECT 
    publishers.publisher,
    COUNT(books.book_id) AS books_cnt
FROM 
    books
    LEFT JOIN publishers ON books.publisher_id = publishers.publisher_id
WHERE 
    num_pages > 50
GROUP BY
    publishers.publisher
ORDER BY
    books_cnt DESC
       '''
publishers_max = pd.io.sql.read_sql(query, con = engine)
display(publishers_max.head())

publisher  books_cnt
0             Penguin Books         42
1                   Vintage         31
2  Grand Central Publishing         25
3          Penguin Classics         24
4          Ballantine Books         19

Издательство, которое выпустило наибольшее число книг толще 50 страниц - Penguin Books. На его счету 42 книги.

<a id='Section_5'></a>
## 5. Автор с самой высокой средней оценкой книг

In [10]:
query = '''
SELECT 
    authors.author,
    AVG(ratings.rating) AS avg_rating
FROM 
    books
    LEFT JOIN authors ON books.author_id = authors.author_id
    LEFT JOIN ratings ON ratings.book_id = books.book_id
WHERE 
    books.book_id IN
                     (
                     SELECT
                          books.book_id
                     FROM
                          books    
                     LEFT JOIN ratings ON ratings.book_id = books.book_id     
                     GROUP BY
                          books.book_id
                     HAVING 
                          COUNT(ratings.rating_id) >= 50
                      )
GROUP BY
    authors.author  
ORDER BY 
    avg_rating DESC
       '''
author_max = pd.io.sql.read_sql(query, con = engine)
display(author_max.head())

author  avg_rating
0         J.K. Rowling/Mary GrandPré    4.287097
1  Markus Zusak/Cao Xuân Việt Khương    4.264151
2                     J.R.R. Tolkien    4.246914
3                  Louisa May Alcott    4.192308
4                       Rick Riordan    4.080645

 Автор с самой высокой средней оценкой книг 4,28 - J.K. Rowling/Mary GrandPré. 

[Вернуться к оглавлению](#Section_0)

<a id='Section_6'></a>
## 6. Среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [13]:
query = '''
SELECT 
    AVG(SUBQ.reviews_id_cnt) AS avg_reviews_id
FROM
    (SELECT
            reviews.username,
            COUNT(reviews.review_id) AS reviews_id_cnt
    FROM
            reviews
    WHERE reviews.username IN
                                (SELECT
                                    ratings.username
                                FROM
                                    ratings
                                    LEFT JOIN books ON books.book_id = ratings.book_id
                                GROUP BY
                                    ratings.username
                                HAVING
                                    COUNT(ratings.rating_id) > 50)
    GROUP BY
            reviews.username) AS SUBQ

       '''
reviews_avg = pd.io.sql.read_sql(query, con = engine)
display(reviews_avg.head())        
        

avg_reviews_id
0       24.333333

 Среднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24.

<a id='Section_7'></a>
## 7. Выводы

1. Количество книг, которые вышли после 1 января 2000 года - 821.

2. Самое большое количество обзоров у книги Twilight (Twilight #1) со средней оценкой 3,66. Есть 6 книг, которые не имеют пользовательских обзоров. 

3. Издательство, которое выпустило наибольшее число книг толще 50 страниц - Penguin Books. На его счету 42 книги. 

4. Автор с самой высокой средней оценкой книг 4,28 - J.K. Rowling/Mary GrandPré. 

5. Среднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24.